sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuAggregateInPandasExec.scala

```diff
@@ -0,0 +1,269 @@
+ *
+ * This plan works by sending the necessary (projected) input grouped data as Arrow record batches
+ * to the python worker, the python worker invokes the UDF and sends the results to the executor,
+ * finally the executor evaluates any post-aggregation expressions and join the result with the
+ * grouped key.
+ *
+ * This node aims at accelerating the data transfer between JVM and Python for GPU pipeline, and
+ * scheduling GPU resources for its Python processes.
+ */
+case class GpuAggregateInPandasExec(
+    groupingExpressions: Seq[NamedExpression],
+    udfExpressions: Seq[GpuPythonUDF],
+    resultExpressions: Seq[NamedExpression],
+    child: SparkPlan)
+  extends UnaryExecNode with GpuPythonExecBase {
+
+  override val output: Seq[Attribute] = resultExpressions.map(_.toAttribute)
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  override def producedAttributes: AttributeSet = AttributeSet(output)
+
+  override def requiredChildDistribution: Seq[Distribution] = {
+    if (groupingExpressions.isEmpty) {
+      AllTuples :: Nil
+    } else {
+      ClusteredDistribution(groupingExpressions) :: Nil
+    }
+  }
+
+  private def collectFunctions(udf: GpuPythonUDF): (ChainedPythonFunctions, Seq[Expression]) = {
+    udf.children match {
+      case Seq(u: GpuPythonUDF) =>
+        val (chained, children) = collectFunctions(u)
+        (ChainedPythonFunctions(chained.funcs ++ Seq(udf.func)), children)
+      case children =>
+        // There should not be any other UDFs, or the children can't be evaluated directly.
+        assert(children.forall(_.find(_.isInstanceOf[GpuPythonUDF]).isEmpty))
+        (ChainedPythonFunctions(Seq(udf.func)), udf.children)
+    }
+  }
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
+    Seq(groupingExpressions.map(ShimLoader.getSparkShims.sortOrder(_, Ascending)))
+
+  // One batch as input to keep the integrity for each group.
+  // (This should be replaced by an iterator that can split batches on key boundaries eventually.
+  //  For more details please refer to the following issue:
+  //        https://github.com/NVIDIA/spark-rapids/issues/2200 )
+  //
+  // But there is a special case that the iterator can not work. when 'groupingExpressions' is
+  // empty, the input must be a single batch, because the whole data is treated as a single
+  // group, and it should be sent to Python at one time.
+  override def childrenCoalesceGoal: Seq[CoalesceGoal] = Seq(RequireSingleBatch)
+
+  // When groupingExpressions is not empty, the input batch will be split into multiple
+  // batches by the grouping expressions, and processed by Python executors group by group,
+  // so better to coalesce the output batches.
+  override def coalesceAfter: Boolean = groupingExpressions.nonEmpty
+
+  override def doExecuteColumnar(): RDD[ColumnarBatch] = {
+    val (mNumInputRows, mNumInputBatches, mNumOutputRows, mNumOutputBatches,
+      spillCallback) = commonGpuMetrics()
```

Comment:
Doesn't need to be handled in this PR, but this many entries in a sequence is better handled with a `case class` to avoid accidentally swizzling the output order on the `commonGpuMetrics` result.

Reply:
got it.